src/places.py

```diff
@@ -384,3 +401,41 @@ def _on_encrypted_eject_finished(self, encrypted, success):
             self._on_device_removed(None, encrypted)
         else:
             self.emit("failed", None)
+
+    def _find_place_by_path(self, listbox, path):
+        for place in listbox:
+            if place.path == path:
+                return place
+        return None
+
+    def _remove_bookmark(self, bookmark, path):
+        place = self._find_place_by_path(self._bookmarks_listbox, path)
+        if place is not None:
+            self._bookmarks_group.remove(place)
+            return True
+        return False
+
+    def _add_bookmark(self, bookmark, path):
+        place = self._find_place_by_path(self._bookmarks_listbox, path)
+        if place is None:
+            self._add_bookmark_place(path)
+            return True
+        return False
+
+    def _add_unadded_accessible_bookmarks(self):
+        for path in self._bookmarks.bookmarked:
+            notAdded = self._find_place_by_path(self._bookmarks_listbox, path) is None
+            if notAdded and os.path.isdir(path): # Might be a bookmarked path on a unmounted device
+                self._add_bookmark_place(path)
```

Comment:
Why do you need to check if it's in widget? If you're worried about duplicates, that's something you can deal with in the mode (e.g., using a set instead of a list). Same thing for available or unavailable bookmarks.
 
So this ideally look like:

```python
for path in self._bookmarks.bookmarked:
    self._add_bookmark_place(path)
```

Reply:
One may connect a device which has paths stored in their bookmarks, the check was to not repeat add to the collection. As on start up it can be come disjointed from the list.